src/mainboard/raptor-cs/talos-2/romstage.c

```diff
@@ -8,39 +8,333 @@
+				       " Marking as nonfunctional.\n", mcs, mca);
+				mark_nonfunctional(mcs, mca);
+				continue;
+			}
+
+			mem_data.mcs[mcs].mca[mca].cl = min;
+
+			/*
+			 * There are also minimal values in Table 170 of JEDEC Standard No. 79-4C which
+			 * probably should also be honored. Some of them (e.g. RRD) depend on the page
+			 * size, which depends on DRAM width. On tested DIMM they are just right - it is
+			 * either minimal legal value or rounded up to whole clock cycle. Can we rely on
+			 * vendors to put sane values in SPD or do we have to check them for validity?
+			 */
+
+			/* Minimum CAS to CAS Delay Time, Same Bank Group */
+			mem_data.mcs[mcs].mca[mca].nccd_l = find_min_mtb_ftb(dimm, 40, 117);
+
+			/* Minimum Write to Read Time, Different Bank Group */
+			mem_data.mcs[mcs].mca[mca].nwtr_s = find_min_multi_mtb(dimm, 44, 43, 0x0F, 8);
+
+			/* Minimum Write to Read Time, Same Bank Group */
+			mem_data.mcs[mcs].mca[mca].nwtr_l = find_min_multi_mtb(dimm, 45, 43, 0xF0, 4);
+
+			/* Minimum Four Activate Window Delay Time */
+			mem_data.mcs[mcs].mca[mca].nfaw = find_min_multi_mtb(dimm, 37, 36, 0x0F, 8);
+
+			/* Minimum RAS to CAS Delay Time */
+			mem_data.mcs[mcs].mca[mca].nrcd = find_min_mtb_ftb(dimm, 25, 122);
+
+			/* Minimum Row Precharge Delay Time */
+			mem_data.mcs[mcs].mca[mca].nrp = find_min_mtb_ftb(dimm, 26, 121);
+
+			/* Minimum Active to Precharge Delay Time */
+			mem_data.mcs[mcs].mca[mca].nras = find_min_multi_mtb(dimm, 28, 27, 0x0F, 8);
+
+			/* Minimum Write Recovery Time */
+			mem_data.mcs[mcs].mca[mca].nwr = find_min_multi_mtb(dimm, 42, 41, 0x0F, 8);
+
+			/* Minimum Activate to Activate Delay Time, Different Bank Group */
+			mem_data.mcs[mcs].mca[mca].nrrd_s = find_min_mtb_ftb(dimm, 38, 119);
+
+			/* Minimum Activate to Activate Delay Time, Same Bank Group */
+			mem_data.mcs[mcs].mca[mca].nrrd_l = find_min_mtb_ftb(dimm, 39, 118);
+
+			/* Minimum Refresh Recovery Delay Time */
+			/* Assuming no fine refresh mode. */
+			mem_data.mcs[mcs].mca[mca].nrfc = find_min_multi_mtb(dimm, 30, 31, 0xFF, 8);
+
+			/* Minimum Refresh Recovery Delay Time for Different Logical Rank (3DS only) */
+			/*
+			 * This one is set per MCA, but it depends on DRAM density, which can be
+			 * mixed between DIMMs under the same channel. We need to choose the bigger
+			 * minimum time, which corresponds to higher density.
+			 *
+			 * Assuming no fine refresh mode.
+			 */
+			val0 = dimm[0].present ? dimm[0].spd[4] & 0xF : 0;
+			val1 = dimm[1].present ? dimm[1].spd[4] & 0xF : 0;
+			min = (val0 < val1) ? val1 : val0;
+
+			switch (min) {
+				case 0x4:
+					mem_data.mcs[mcs].mca[mca].nrfc_dlr = ns_to_nck(90);
+					break;
+				case 0x5:
+					mem_data.mcs[mcs].mca[mca].nrfc_dlr = ns_to_nck(120);
+					break;
+				case 0x6:
+					mem_data.mcs[mcs].mca[mca].nrfc_dlr = ns_to_nck(185);
+					break;
+				default:
+					die("Unsupported DRAM density\n");
+			}
```

Comment:
The code from lines 209 to 300 could be a function. This kind of nested level is hard to follow when a for loop is so long